apps/api/e2e/utils/server.ts

```diff
@@ -18,167 +18,64 @@ const BASE_URL = `http://${HOST}:${PORT}${API_PREFIX ? `/${API_PREFIX}` : ""}`;
-          process.kill(-serverProcess.pid, "SIGTERM");
-          serverProcess = null;
-        } catch (error) {
-          console.error("Error killing server process:", error);
-        }
-      }
+  console.log("\n\n\n\n");
+  console.log("server process:", serverProcess.pid);
+  console.log("\n\n\n\n");
 
-      // Close the server
-      server.close(() => {
-        console.log("Server stopped");
-        resolve();
-      });
-    } catch (error) {
-      console.error("Error in stopServer:", error);
-      resolve(); // Resolve anyway to avoid hanging promises
-    }
-  });
+  // Wait for the server to be ready
+  return await waitForServerReady(30, 500); // 30 retries, 500ms interval = 15 seconds max
 }
 
-/**
- * Kill any existing server processes running on the test ports
- * This includes both the main API server and the metrics server
- */
-export async function killExistingServers(): Promise<void> {
-  return new Promise<void>((resolve) => {
+export function stopServer() {
+  // Kill the server process if it exists
+  if (serverProcess && serverProcess.pid) {
     try {
-      const testPort = process.env.TEST_PORT || "3001";
-      const metricsPort = process.env.METRICS_PORT || "3003";
-      console.log(
-        `Checking for existing servers on ports ${testPort} and ${metricsPort}...`,
-      );
-
-      // Platform-specific command to find and kill processes using both ports
-      let command: string;
-      let args: string[];
-
-      if (process.platform === "win32") {
-        // Windows - kill processes on both ports
-        command = "cmd.exe";
-        args = [
-          "/c",
-          `(for /f "tokens=5" %a in ('netstat -ano ^| findstr :${testPort}') do taskkill /F /PID %a) & (for /f "tokens=5" %a in ('netstat -ano ^| findstr :${metricsPort}') do taskkill /F /PID %a)`,
-        ];
-      } else {
-        // Unix-like (macOS, Linux) - kill processes on both ports
-        command = "bash";
-        args = [
-          "-c",
-          `(lsof -i:${testPort} -t | xargs -r kill -9); (lsof -i:${metricsPort} -t | xargs -r kill -9)`,
-        ];
+      console.log(`Killing server process with PID: ${serverProcess.pid}`);
+      // Negative PID to kill process group, i.e all child processes
+      kill(serverProcess.pid);
+      serverProcess = null;
+    } catch (error: unknown) {
+      // If there is not a process with the expected PID we don't have to do
+      // anything since it's already been killed. This allows calling this
+      // function repeatedly.
+      if (
+        !(
+          typeof error === "object" &&
+          error !== null &&
+          "code" in error &&
+          error.code === "ESRCH"
+        )
+      ) {
+        // If there is some other error, we can log it and throw.
+        console.error("Error killing server process:", error);
+        throw error;
       }
-
-      const killProcess = spawn(command, args, { stdio: "pipe" });
```

Comment:
We have been blindly killing processes based only on the port they use.  This is quite dangerous if something important happened to be using port 3001.